File: src/main/java/com/refinedmods/refinedstorage/network/grid/GridItemDeltaMessage.java
package com.refinedmods.refinedstorage.network.grid;

import com.refinedmods.refinedstorage.RS;
import com.refinedmods.refinedstorage.screen.BaseScreen;
import com.refinedmods.refinedstorage.screen.grid.GridScreen;
import com.refinedmods.refinedstorage.screen.grid.stack.ItemGridStack;
import com.refinedmods.refinedstorage.util.StackUtils;

import java.util.LinkedList;
import java.util.List;

import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import net.neoforged.neoforge.network.handling.PlayPayloadContext;

public class GridItemDeltaMessage implements CustomPacketPayload {
    public static final ResourceLocation ID = new ResourceLocation(RS.ID, "grid_item_delta");

    private final List<GridStackDelta<ItemGridStack>> deltas;

    public GridItemDeltaMessage(List<GridStackDelta<ItemGridStack>> deltas) {
        this.deltas = deltas;
    }

    public static GridItemDeltaMessage decode(FriendlyByteBuf buf) {
        final int size = buf.readInt();
        final List<GridStackDelta<ItemGridStack>> deltas = new LinkedList<>();
        for (int i = 0; i < size; ++i) {
            deltas.add(StackUtils.readItemGridStackDelta(buf));
        }
        return new GridItemDeltaMessage(deltas);
    }

    public static void handle(GridItemDeltaMessage message, PlayPayloadContext ctx) {
        BaseScreen.executeLater(GridScreen.class, grid -> {
            message.deltas.forEach(p -> grid.getView().postChange(p.stack(), p.change()));
        });
    }

    @Override
    public void write(FriendlyByteBuf buf) {
        buf.writeInt(deltas.size());
        deltas.forEach(delta -> StackUtils.writeItemGridStackDelta(buf, delta));
    }

    @Override
    public ResourceLocation id() {
        return ID;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/network/grid/GridItemDeltaMessage.kt
package com.refinedmods.refinedstorage.network.grid

import com.refinedmods.refinedstorage.api.network.INetwork
import com.refinedmods.refinedstorage.api.util.IComparer
import com.refinedmods.refinedstorage.api.util.StackListEntry
import com.refinedmods.refinedstorage.api.util.StackListResult
import com.refinedmods.refinedstorage.screen.BaseScreen
import com.refinedmods.refinedstorage.screen.grid.GridScreen
import com.refinedmods.refinedstorage.screen.grid.stack.IGridStack
import com.refinedmods.refinedstorage.util.StackUtils.readItemGridStack
import com.refinedmods.refinedstorage.util.StackUtils.writeItemGridStack
import net.minecraft.item.ItemStack
import net.minecraft.network.PacketByteBuf
import net.minecraftforge.fml.network.NetworkEvent
import org.apache.commons.lang3.tuple.Pair
import java.util.*
import java.util.function.Consumer
import java.util.function.Supplier

class GridItemDeltaMessage {
    @Nullable
    private var network: INetwork? = null
    private var deltas: List<StackListResult<ItemStack>>? = null
    private var clientDeltas: List<Pair<IGridStack, Int>>? = null

    constructor(network: INetwork?, deltas: List<StackListResult<ItemStack>>?) {
        this.network = network
        this.deltas = deltas
    }

    constructor(clientDeltas: List<Pair<IGridStack, Int>>?) {
        this.clientDeltas = clientDeltas
    }

    companion object {
        fun decode(buf: PacketByteBuf): GridItemDeltaMessage {
            val size: Int = buf.readInt()
            val clientDeltas: MutableList<Pair<IGridStack, Int>> = LinkedList()
            for (i in 0 until size) {
                val delta: Int = buf.readInt()
                clientDeltas.add(Pair.of(readItemGridStack(buf), delta))
            }
            return GridItemDeltaMessage(clientDeltas)
        }

        fun encode(message: GridItemDeltaMessage, buf: PacketByteBuf) {
            buf.writeInt(message.deltas!!.size)
            for (delta in message.deltas!!) {
                buf.writeInt(delta.change)
                val craftingEntry: StackListEntry<ItemStack>? = message.network!!.itemStorageCache!!.getCraftablesList()!!.getEntry(delta.stack, IComparer.COMPARE_NBT)
                writeItemGridStack(buf, delta.stack, delta.id, craftingEntry?.id, false, message.network!!.itemStorageTracker!![delta.stack])
            }
        }

        fun handle(message: GridItemDeltaMessage, ctx: Supplier<NetworkEvent.Context>) {
            BaseScreen.executeLater(GridScreen::class.java) { grid: GridScreen ->
                message.clientDeltas!!.forEach(Consumer { p: Pair<IGridStack, Int> -> grid.view.postChange(p.left, p.right) })
                grid.view.sort()
            }
            ctx.get().setPacketHandled(true)
        }
    }
}